prow/metrics/prowjobs/collector.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prowjobs
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/tools/cache"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+func update(histogramVec *prometheus.HistogramVec, oldJob *prowapi.ProwJob, newJob *prowapi.ProwJob) {
+	if oldJob == nil {
+		return
+	}
+	if oldJob.Status.State == newJob.Status.State {
+		return
+	}
+
+	var oldTime *metav1.Time
+	var newTime *metav1.Time
+
+	switch oldJob.Status.State {
+	case prowapi.TriggeredState:
+		oldTime = &oldJob.CreationTimestamp
+	case prowapi.PendingState:
+		if oldJob.Status.PendingTime == nil {
```

Comment:
I still think it would be way clearer to just do this check once ... do both of the `switch` statements, then if either is `nil`, return. That way we don't seven different places we need to `return`

Reply:
ok, let me look into that.

Reply:
Example was here https://github.com/kubernetes/test-infra/pull/14695#discussion_r334089698

Reply:
and you were definitely right. Done.

Reply:
> Example was here [#14695 (comment)](https://github.com/kubernetes/test-infra/pull/14695#discussion_r334089698)

Yes, sorry missed it.